vecindex: add more logic tests for vector indexes

andy-kimball opened the pull request:
Add more logic tests for vector indexes, in order to expand our testing coverage, particularly when it comes to partial indexes.

Fix a buglet where fixups in a deterministic index weren't being suspended. Make it so that a deterministic index always automatically suspends its fixups, without needing the caller to do so.

Epic: CRDB-42943
Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/146045)
<!-- Reviewable:end -->

andy-kimball committed:
vecindex: add more logic tests for vector indexes

Add more logic tests for vector indexes, in order to expand our testing
coverage, particularly when it comes to partial indexes.

Fix a buglet where fixups in a deterministic index weren't being suspended.
Make it so that a deterministic index always automatically suspends its
fixups, without needing the caller to do so.

Epic: CRDB-42943
Release note: None

DrewKimball reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 9 of 9 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/146045#-:-OPXndVL9BTUdGjTlu1t:bnfp4nl)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @mw5h)

___
*[`pkg/sql/opt/exec/execbuilder/testdata/partial_index` line 1099 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/146045#-OPXjZehDF5ulbzMy9-K:-OPXjZehDF5ulbzMy9-L:b-rsrz0j) ([raw file](https://github.com/cockroachdb/cockroach/blob/159be7eca754e026fc03dca6d0653a7d0948baa6/pkg/sql/opt/exec/execbuilder/testdata/partial_index#L1099)):*
> ```
> SELECT a FROM vec@i WHERE c IN ('foo', 'bar') ORDER BY b <-> '[1,2]' LIMIT 10
> ----
> Scan /Table/20/1/10{8-9}
> ```

Does kvtrace only show requests with non-empty results? Why don't we see scans conducted internally to the vector index?
___
*[`pkg/sql/logictest/testdata/logic_test/partial_index` line 1402 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/146045#-OPTu5f6AWttfF4jkxXb:-OPTu5f6AWttfF4jkxXc:b-j8wumq) ([raw file](https://github.com/cockroachdb/cockroach/blob/159be7eca754e026fc03dca6d0653a7d0948baa6/pkg/sql/logictest/testdata/logic_test/partial_index#L1402)):*
> ```
> statement ok
> CREATE TABLE vec (k INT PRIMARY KEY, v VECTOR(2), s STRING);
> CREATE VECTOR INDEX i ON vec (s, v) WHERE s IN ('foo', 'bar');
> ```

It might be interesting to have a partial index without the prefix, or a different prefix column than the one in the index predicate.


<!-- Sent from Reviewable.io -->

andy-kimball reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/146045)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @DrewKimball and @mw5h)

___
*[`pkg/sql/logictest/testdata/logic_test/partial_index` line 1402 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/146045#-OPTu5f6AWttfF4jkxXb:-OPYRtld79sZ112RIfGV:b-3cmly3) ([raw file](https://github.com/cockroachdb/cockroach/blob/159be7eca754e026fc03dca6d0653a7d0948baa6/pkg/sql/logictest/testdata/logic_test/partial_index#L1402)):*
<details><summary><i>Previously, DrewKimball (Drew Kimball) wrote…</i></summary><blockquote>

It might be interesting to have a partial index without the prefix, or a different prefix column than the one in the index predicate.
</blockquote></details>

How do we have a partial vector index without the prefix? The prefix columns are the only columns that can be filtered by the partial index predicate. If we use other columns that are not part of the vector index, then how does that test the vector index?
___
*[`pkg/sql/opt/exec/execbuilder/testdata/partial_index` line 1099 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/146045#-OPXjZehDF5ulbzMy9-K:-OPYSdYd2jS6n1zhEdvV:b1fpjz4) ([raw file](https://github.com/cockroachdb/cockroach/blob/159be7eca754e026fc03dca6d0653a7d0948baa6/pkg/sql/opt/exec/execbuilder/testdata/partial_index#L1099)):*
<details><summary><i>Previously, DrewKimball (Drew Kimball) wrote…</i></summary><blockquote>

Does kvtrace only show requests with non-empty results? Why don't we see scans conducted internally to the vector index?
</blockquote></details>

There aren't any rows in the index at this point, and just the root level. What other scans were you expecting?


<!-- Sent from Reviewable.io -->

DrewKimball reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/146045)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @andy-kimball and @mw5h)

___
*[`pkg/sql/logictest/testdata/logic_test/partial_index` line 1402 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/146045#-OPTu5f6AWttfF4jkxXb:-OPcZjwKE7M0mScfPfeL:bv8m9od) ([raw file](https://github.com/cockroachdb/cockroach/blob/159be7eca754e026fc03dca6d0653a7d0948baa6/pkg/sql/logictest/testdata/logic_test/partial_index#L1402)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

How do we have a partial vector index without the prefix? The prefix columns are the only columns that can be filtered by the partial index predicate. If we use other columns that are not part of the vector index, then how does that test the vector index?
</blockquote></details>

Hm, I see the problem. We should probably be able to plan a partial-index-scan when the filters exactly match the predicate, but currently that's not the case. Disregard my original comment for now.
___
*[`pkg/sql/opt/exec/execbuilder/testdata/partial_index` line 1099 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/146045#-OPXjZehDF5ulbzMy9-K:-OPcWsGI84uquhezNZy8:bpfb8ad) ([raw file](https://github.com/cockroachdb/cockroach/blob/159be7eca754e026fc03dca6d0653a7d0948baa6/pkg/sql/opt/exec/execbuilder/testdata/partial_index#L1099)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

There aren't any rows in the index at this point, and just the root level. What other scans were you expecting?
</blockquote></details>

I'd expect to see the root level scan. the one printed out here is for the namespace table, or something, during planning.


<!-- Sent from Reviewable.io -->

andy-kimball reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/146045)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @DrewKimball and @mw5h)

___
*[`pkg/sql/opt/exec/execbuilder/testdata/partial_index` line 1099 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/146045#-OPXjZehDF5ulbzMy9-K:-OPccl7V9IhK_MC9xi1y:bwkjepo) ([raw file](https://github.com/cockroachdb/cockroach/blob/159be7eca754e026fc03dca6d0653a7d0948baa6/pkg/sql/opt/exec/execbuilder/testdata/partial_index#L1099)):*
<details><summary><i>Previously, DrewKimball (Drew Kimball) wrote…</i></summary><blockquote>

I'd expect to see the root level scan. the one printed out here is for the namespace table, or something, during planning.
</blockquote></details>

Yes, you're right, it must be getting filtered out, not sure how I didn't notice it. I added a new logging statement in `store_txn.go` that will show the `Scan` that we do.


<!-- Sent from Reviewable.io -->

andy-kimball commented:
*[`pkg/sql/logictest/testdata/logic_test/partial_index` line 1402 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/146045#-OPTu5f6AWttfF4jkxXb:-OPcok_63aGmUIO5Mn4A:ba4taz4) ([raw file](https://github.com/cockroachdb/cockroach/blob/159be7eca754e026fc03dca6d0653a7d0948baa6/pkg/sql/logictest/testdata/logic_test/partial_index#L1402)):*
<details><summary><i>Previously, DrewKimball (Drew Kimball) wrote…</i></summary><blockquote>

Hm, I see the problem. We should probably be able to plan a partial-index-scan when the filters exactly match the predicate, but currently that's not the case. Disregard my original comment for now.
</blockquote></details>

I'm not sure I understand...we are planning partial-index scans when the filters exactly match, e.g.:
```
root@localhost:26257/defaultdb> explain SELECT * FROM vec@i WHERE s = 'foo' ORDER BY v <-> '[5,5]' LIMIT 1;
                      info
-------------------------------------------------
  distribution: local
  vectorized: false

  • top-k
  │ estimated row count: 1
  │ order: +column8
  │ k: 1
  │
  └── • render
      │
      └── • lookup join
          │ table: vec@vec_pkey
          │ equality: (k) = (k)
          │ equality cols are key
          │
          └── • vector search
                table: vec@i (partial index)
                target count: 1
                prefix spans: [/'foo' - /'foo']
```


<!-- Sent from Reviewable.io -->

DrewKimball reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/146045)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @andy-kimball and @mw5h)

___
*[`pkg/sql/logictest/testdata/logic_test/partial_index` line 1402 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/146045#-OPTu5f6AWttfF4jkxXb:-OPcyguXCGqjPOw9DOeL:bfzsvkx) ([raw file](https://github.com/cockroachdb/cockroach/blob/159be7eca754e026fc03dca6d0653a7d0948baa6/pkg/sql/logictest/testdata/logic_test/partial_index#L1402)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

I'm not sure I understand...we are planning partial-index scans when the filters exactly match, e.g.:
```
root@localhost:26257/defaultdb> explain SELECT * FROM vec@i WHERE s = 'foo' ORDER BY v <-> '[5,5]' LIMIT 1;
                      info
-------------------------------------------------
  distribution: local
  vectorized: false

  • top-k
  │ estimated row count: 1
  │ order: +column8
  │ k: 1
  │
  └── • render
      │
      └── • lookup join
          │ table: vec@vec_pkey
          │ equality: (k) = (k)
          │ equality cols are key
          │
          └── • vector search
                table: vec@i (partial index)
                target count: 1
                prefix spans: [/'foo' - /'foo']
```
</blockquote></details>

Yes, I mean when there's no prefix column.


<!-- Sent from Reviewable.io -->

DrewKimball reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/146045)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @andy-kimball and @mw5h)

___
*[`pkg/sql/logictest/testdata/logic_test/partial_index` line 1402 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/146045#-OPTu5f6AWttfF4jkxXb:-OPd79GHAxFia_RnbGy4:b-1puh9u) ([raw file](https://github.com/cockroachdb/cockroach/blob/159be7eca754e026fc03dca6d0653a7d0948baa6/pkg/sql/logictest/testdata/logic_test/partial_index#L1402)):*
<details><summary><i>Previously, DrewKimball (Drew Kimball) wrote…</i></summary><blockquote>

Yes, I mean when there's no prefix column.
</blockquote></details>

I discovered https://github.com/cockroachdb/cockroach/issues/146257 and https://github.com/cockroachdb/cockroach/issues/146258 based on this, will fix soon.


<!-- Sent from Reviewable.io -->

andy-kimball commented:
bors r=drewkimball
